zipline/finance/performance/position_tracker.py

```diff
@@ -64,6 +69,71 @@ def _update_multipliers(self, sid):
                     asset.contract_multiplier
                 self._position_payout_multipliers[sid] = \
                     asset.contract_multiplier
+                # Futures are closed on their notice_date
+                if asset.notice_date:
+                    self._insert_auto_close_position_date(
+                        dt=asset.notice_date,
+                        sid=sid
+                    )
+                # If the Future does not have a notice_date, it will be closed
+                # on its expiration_date
+                elif asset.expiration_date:
+                    self._insert_auto_close_position_date(
+                        dt=asset.expiration_date,
+                        sid=sid
+                    )
+
+    def _insert_auto_close_position_date(self, dt, sid):
+        """
+        Inserts the given SID in to the list of positions to be auto-closed by
+        the given dt.
+
+        Parameters
+        ----------
+        dt : pandas.Timestamp
+            The date before-which the given SID will be auto-closed
+        sid : int
+            The SID of the Asset to be auto-closed
+        """
+        date_sids = self.auto_close_position_dates.setdefault(dt, [])
+        if sid not in date_sids:
+            date_sids.append(sid)
+
+    def auto_close_position_events(self, next_trading_day):
+        """
+        Generates CLOSE_POSITION events for any SIDs whose auto-close date is
+        before or equal to the given date.
+
+        Parameters
+        ----------
+        next_trading_day : pandas.Timestamp
+            The time before-which certain Assets need to be closed
+
+        Yields
+        ------
+        Event
+            A close position event for any sids that should be closed before
+            the next_trading_day parameter
+        """
+        past_asset_end_dates = set()
+
+        # Check the auto_close_position_dates dict for SIDs to close
+        for date, sids in self.auto_close_position_dates.items():
```

Comment:
No, but one potential optimization is to pre-sort the auto-close dates on insertion, such that we can bail out of the generator as soon as we hit the first date after next_trading_day. What do you think?